### PR TITLE
feat: implement assignment of priorities to Kong routes from GRPCRoutes

### LINKED ISSUE #4250: Assign priorities to Kong expression routes from GRPCRoutes
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem Statement

Separate of https://github.com/Kong/kubernetes-ingress-controller/issues/3958. The priorities of routes translated from GRPCRoutes should follow the specification of the doc: https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1alpha2.GRPCRouteRule.

### Proposed Solution

_No response_

### Additional information

_No response_

### Acceptance Criteria

- [ ] priorities of routes translated from GRPCRoutes satisfies the specification of [doc](https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1alpha2.GRPCRouteRule)
- [ ]  priorities of the routes do not interleave with routes translated from other kinds of resources

### PR DESCRIPTION
**Pull request by randmonkey**
**What this PR does / why we need it**:

Implement assigning priorities to Kong routes translated from `GRPCRoute`.

**Which issue this PR fixes**:

fixes #4250 

**Special notes for your reviewer**:

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [x] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR

### CODE REVIEW

**Review by rainest — CHANGES_REQUESTED** [review]
Making a set of complete GRPCRoute objects that obey additional rules about their rule counts (and validating those rules downstream) feels a bit odd. I feel like we may get clearer code if we create new structs that _must_ obey those rules and has dedicated fields for the rule/match indices instead of adding annotations to the copies. Something like

```
type SplitRuleCollection struct {
	source *gatewayv1alpha2.GRPCRout

	rulesByHostname map[string]Rule
}

// Rule matches the structure of GRPRouteRule, but uses a modified Match type,
// so that we can loop over the Match slice and assign priority correctly,
// and then apply the filters and backends to the generated Kong routes.

type MatchForPriority struct {
	Match gatewayv1alpha2.GRPCRouteMatch
	RuleIndex int
	MatchIndex int
}
```

This has a fair amount of additional validation and conversion code that it seems like we could avoid with dedicated types. Dedicated types would also make it obvious when you're dealing with the derived type or not, whereas as-is it's not obvious in the code whether you're looking at a split GRPCRoute or a regular one.

However, doing the above would require refactoring just about everything in the PR, so I don't think I'd request it unless there's broader consensus from other reviewers that reusing the original type makes the code confusing enough to warrant the change.

File: internal/dataplane/parser/translators/grpcroute_atc.go [hunk 1]
```diff
@@ -169,3 +172,327 @@ func getGRPCRouteHostnamesAsSliceOfStrings(grpcroute *gatewayv1alpha2.GRPCRoute)
...
+// Where:
+// P: set to 1 if the hostname is non-wildcard.
+// host len: length of hostname.
+// GRPC service length: length of `Service` part in method match
+// GRPC method length: length of `Method` part in method match
+// Header No.: number of header matches.
+// relative order: relative order of creation timestamp, namespace and name and internal rule/match order between different (split) GRPCRoutes.
+//
+// REVIEW: althogh not specified in official docs, do we need to assign a bit for GRPC method match type
+// to assign higher priority for method match with `Exact` match?
```
[the remark this answers is not in the document]
Comment (randmonkey) [reply internal/dataplane/parser/translators/grpcroute_atc.go:292]: https://github.com/kubernetes-sigs/gateway-api/issues/2216 Created an upstream issue to require clarification here. Before the question gets answered, I think it is necessary to preserve a bit for that.

**Comment by randmonkey** [comment]
> Making a set of complete GRPCRoute objects that obey additional rules about their rule counts (and validating those rules downstream) feels a bit odd. I feel like we may get clearer code if we create new structs that _must_ obey those rules and has dedicated fields for the rule/match indices instead of adding annotations to the copies. Something like
> 
> ```
> type SplitRuleCollection struct {
> 	source *gatewayv1alpha2.GRPCRout
> 
> 	rulesByHostname map[string]Rule
> }
> 
> // Rule matches the structure of GRPRouteRule, but uses a modified Match type,
> // so that we can loop over the Match slice and assign priority correctly,
> // and then apply the filters and backends to the generated Kong routes.
> 
> type MatchForPriority struct {
> 	Match gatewayv1alpha2.GRPCRouteMatch
> 	RuleIndex int
> 	MatchIndex int
> }
> ```
> 
> This has a fair amount of additional validation and conversion code that it seems like we could avoid with dedicated types. Dedicated types would also make it obvious when you're dealing with the derived type or not, whereas as-is it's not obvious in the code whether you're looking at a split GRPCRoute or a regular one.
> 
> However, doing the above would require refactoring just about everything in the PR, so I don't think I'd request it unless there's broader consensus from other reviewers that reusing the original type makes the code confusing enough to warrant the change.

:+1:  I will do a refactor to remove the direct usage of `GRPCRoute` in the split rules. The assigning of priorities to routes from  `HTTPRoute`s had been also implemented in this way, would you please create another issue for refactoring that after this PR merged? @rainest

File: internal/dataplane/parser/translators/grpcroute_atc.go [hunk 2]
```diff
@@ -169,3 +172,327 @@ func getGRPCRouteHostnamesAsSliceOfStrings(grpcroute *gatewayv1alpha2.GRPCRoute)
...
+		hostnameInRouteName,
+		grpcRoute.Annotations[InternalRuleIndexAnnotationKey],
+		grpcRoute.Annotations[InternalMatchIndexAnnotationKey],
+	)
+
+	r := kongstate.Route{
+		Route: kong.Route{
+			Name:         kong.String(routeName),
+			PreserveHost: kong.Bool(true),
+			// TODO: add `StripPath: false` here?
```
[the remark this answers is not in the document]
Comment (randmonkey) [reply internal/dataplane/parser/translators/grpcroute_atc.go]: removed the TODO.

File: CHANGELOG.md [hunk 3]
```diff
@@ -108,6 +108,10 @@ Adding a new version? You'll need three changes:
   them to expression based Kong routes. The assigning method follows the
   [specification on priorities of matches in `HTTPRoute`][httproute-specification].
   [#4296](https://github.com/Kong/kubernetes-ingress-controller/pull/4296)
+- Assign priorities to routes translated from GRPCRoutes when parser translates 
+  them expression based Kong routes. The assigning method follows the
+  [specification on priorities of matches in `GRPCRoute`][grpcroute-specification].
+  [#4364](https://github.com/Kong/kubernetes-ingress-controller/pull/4364)
```
[the remark this answers is not in the document]
Comment (randmonkey) [reply CHANGELOG.md]: done.

File: internal/dataplane/parser/translate_grpcroute.go [hunk 4]
```diff
@@ -79,6 +85,73 @@ func (p *Parser) ingressRulesFromGRPCRoute(result *ingressRules, grpcroute *gate
 	return nil
 }
 
+// ingressRulesFromGRPCRoutesUsingExpressionRoutes translates GRPCRoutes to expression based routes
+// when ExpressionRoutes feature flag is enabled.
+// Because we need to assign different priorities based on the hostname and match in the specification of HTTPRoutes,
```
[the remark this answers is not in the document]
Comment (randmonkey) [reply internal/dataplane/parser/translate_grpcroute.go]: done.

File: internal/dataplane/parser/translators/grpcroute_atc.go [hunk 5]
```diff
@@ -169,3 +172,327 @@ func getGRPCRouteHostnamesAsSliceOfStrings(grpcroute *gatewayv1alpha2.GRPCRoute)
...
+// If ties exists in the first step, where multiple GRPCRoutes has the same priority
+// calculated from the fields, we run a sort for the GRPCRoutes in the tie
+// and assign the bits for "relative order" according to the sorting result of these GRPCRoutes.
+func AssignRoutePriorityToSplitGRPCRoutes(
+	logger logr.Logger,
+	splitGRPCoutes []*gatewayv1alpha2.GRPCRoute,
+) []SplitGRPCRouteToKongRoutePriority {
+	priorityToSplitGRPCRoutes := map[int][]*gatewayv1alpha2.GRPCRoute{}
+	for _, grpcRoute := range splitGRPCoutes {
+		// skip if GRPCRoute does not contain the annotation, because this means the GRPCRoute is not a split one.
```
[the remark this answers is not in the document]
Comment (randmonkey) [reply internal/dataplane/parser/translators/grpcroute_atc.go]: refactored this.

File: CHANGELOG.md [hunk 6]
```diff
@@ -131,6 +135,7 @@ Adding a new version? You'll need three changes:
 
 [gojson]: https://github.com/goccy/go-json
 [httproute-specification]: https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1beta1.HTTPRoute
+[grpcroute-specification]: https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1alpha2.GRPCRoute
```
[the remark this answers is not in the document]
Comment (randmonkey) [reply CHANGELOG.md]: done.

**Review by rainest — APPROVED** [review]

**Pull request merged**